### PR TITLE
SDK export functionalty

### PR DESCRIPTION
**Pull request by eladyitz**

### CODE REVIEW

**Comment by eladyitz** [comment]
the PR is very large so I want to point out the main files:

1. src/sdk.js
- this is the main logic file containing the functionality taken from com.sap.watt.
- I did not change much of the code except bugs that I found and API. 

2. src/utils/*
- all the files there were copied "as-is" from com.sap.watt

3. src/api.js and api.d.ts
- only files dealing with API

4. package.json
- the only change is an adding of:
    "tmp": "0.0.33",
    "sinon": "7.0.0",
    "sinon-chai": "3.2.0"
to devDependencies

5. test/sdk/*
- all the test resources, configuration and the spec file

**Comment by bd82** [comment]
You need to sign the CLA

**Comment by eladyitz** [comment]
I signed the CLA several times

**Comment by bd82** [comment]
Look above:

> Elad Yitzhaik seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please add the email address used for this commit to your account.
You have signed the CLA already but the status is still pending? Let us recheck it.

I think you need to configure your email address in your git client
and make sure that email address is also added to github.com account

**Comment by bd82** [comment]
Your build is failing.
For some reason it is not reported here as a status check.
Perhaps due to issues with your github account?

I will create a PR to check this.

**Comment by bd82** [comment]
I fixed the CI. may need to re-trigger this to get the status check listed

**Comment by bd82** [comment]
O.k. you get status reports now from circle-ci.
I suggest you rebase with master to get the "fixed" CI.
It seems you need to format some source code to comply with prettier formatting.

File: api.d.ts [hunk 1]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
 
 export declare const bundling: BundlingAPI
 
+// ----------------------------------------------------------------
+// -------------------------- SDK ----------------------------
+// ----------------------------------------------------------------
+
+export interface SdkAPI {
```
Comment (bd82) [inline api.d.ts]: I think least 2-3 people should review the APIs and confirm that those have been reviewed

File: api.d.ts [hunk 2]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+  /**
+   * create SDK json files
+   */
+  generateSdkDocumentation: (
```
Comment (bd82) [inline api.d.ts]: For discussion: generateSdkDocumentation vs generateSdk**Docs**

File: api.d.ts [hunk 3]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+    target: string,
+    options?: {
+      /**
+       * Output directory for the sdk.
```
Comment (bd82) [inline api.d.ts]: > Output directory for the sdk.

> Output directory for the sdk **artifacts**.

?

File: api.d.ts [hunk 4]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+      outDir?: string
+
+      /**
+       * Source files to create the sdk from.
+       * Default value is "\w+\.json".
+       */
+      sourcesExcludePattern?: string
+
+      /**
+       * example files to create the sdk from.
```
Comment (bd82) [inline api.d.ts]: upper case "E" in word **example**

File: api.d.ts [hunk 5]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+    target: string,
+    options?: {
+      /**
+       * Output directory for the sdk.
+       * Default value is "dist".
+       */
+      outDir?: string
+
+      /**
+       * Source files to create the sdk from.
```
Comment (bd82) [inline api.d.ts]: The comment here talks about inclusion but the config key is about exclusion.
What are we excluding and why? what use case does this config option serve?

File: api.d.ts [hunk 6]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+// ----------------------------------------------------------------
+// -------------------------- SDK ----------------------------
+// ----------------------------------------------------------------
+
+export interface SdkAPI {
+  /**
+   * create SDK json files
+   */
+  generateSdkDocumentation: (
+    target: string,
```
Comment (bd82) [inline api.d.ts]: Maybe a comment that target is the path to the package.json of the extension/feature?

File: api.d.ts [hunk 7]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+       * Source files to create the sdk from.
+       * Default value is "\w+\.json".
+       */
+      sourcesExcludePattern?: string
+
+      /**
+       * example files to create the sdk from.
+       * Default value is "/.+\/test\/\w+.js/".
+       */
+      examplesIncludePattern?: string
```
Comment (bd82) [inline api.d.ts]: again what are these examples? are they the tests we extract snippets from?

File: api.d.ts [hunk 8]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+
+      /**
+       * Sources directory for the sdk.
+       * Default value is "src".
```
Comment (bd82) [inline api.d.ts]: is the default value "src'? or is it [PATH_TO_target]/src?

File: api.d.ts [hunk 9]
```diff
@@ -186,6 +186,53 @@ export interface BundlingAPI {
...
+       */
+      sourcesDir?: string
+
+      /**
+       * Examples directory for the sdk.
```
Comment (bd82) [inline api.d.ts]: "examples for code snippets"?
again its relative to the target no?

File: src/api.js [hunk 10]
```diff
@@ -2,18 +2,23 @@
...
+const _ = require("lodash")
 
 const bundling = require("./bundling")
 const karma = require("./karma")
 const devServer = require("./dev_server")
+const sdk = require("./sdk")
 
 const API = {}
+const sdkClone = _.cloneDeep(sdk)
+delete sdkClone.test
```
Comment (bd82) [inline src/api.js]: add a comment that we are sanitizing the public API.
and preventing th exposure of "internal" methods

File: src/utils/jsDoc/README.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#JSDoc configuration and extensions
```
Comment (bd82) [inline src/utils/jsDoc/README.md]: Remove these docs, you should not reference internal SAP artifacts in OSS docs

File: src/utils/jsDoc/config.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
+{
+  "plugins": [
+    "C:\\Git\\webide-client-tools\\src\\utils\\jsDoc\\lib\\sapui5-jsdoc3-plugin\\sapui5-jsdoc3.js"
```
Comment (bd82) [inline src/utils/jsDoc/config.json]: what are these absolute windows style paths doing here?

File: src/utils/jsDoc/lib/sapui5-jsdoc3-template/ServiceExample.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
+define(["sap.watt.common.platform/service/ui/AbstractPart"], function(
+  AbstractPart
+) {
```
Comment (bd82) [inline src/utils/jsDoc/lib/sapui5-jsdoc3-template/ServiceExample.js]: remove all sapui5-jsdoc3-template from here!

**Pull request closed without merging**